File: 2021/day01/solution.py
from typing import Iterable

def getLines(filename="input.txt") -> Iterable[str]:
    with open(filename) as f:
        line = f.readline()
        while line:
            yield line[:-1]
            line = f.readline()

def main():
    print(f"part1: {part1()}")
    print(f"part2: {part2()}")

def part1():
    lines = getLines()
    nums = map(lambda x: int(x), lines)
    prev = next(nums)
    increases = 0
    for num in nums:
        # print(f"num={num}, prev={prev}")
        if num > prev:
            increases += 1
        prev = num
    return f"increases={increases}"

def part2():
    lines = getLines()
    nums = map(lambda x: int(x), lines)

    # maintain a 3 number (temporarily 4) window and just compare the first and fourth to determine increase
    from collections import deque
    window = deque()
    while len(window) < 3:
        window.append(next(nums))
    increases = 0
    for num in nums:
        if window.popleft() < num:
            increases += 1
        window.append(num)
    return f"increases={increases}"

if __name__ == "__main__":
    main()

File: 2021/day01/golang/solution.go
package main

import (
    "bufio"
    //"container/list"
    "log"
    "os"
    "strconv"
)


func check(err error) {
    if err != nil {
        log.Fatal(err)
    }
}

func getLines(path string) ([]string, error) {
    file, err := os.Open(path)
    defer file.Close()

    if err != nil {
        return nil, err
    }

    scanner := bufio.NewScanner(file)
    lines := make([]string, 0, 3000)
    for scanner.Scan() {
        line := scanner.Text()
        lines = append(lines, line)
    }

    return lines, nil
}

func main() {
    log.SetFlags(log.Lmicroseconds)
    part1("input.txt")
    part2("input.txt")
}

func part1(path string) {
    log.SetPrefix("part1: ")
    lines, err := getLines(path)
    check(err)

    nums := make([]int, 0, len(lines))
    for _, line := range lines {
        num, err := strconv.Atoi(line)
        check(err)
        nums = append(nums, num)
    }
    var prev *int = nil
    increases := 0
    for i, num := range nums {
        if prev != nil && *prev < num {
            increases++
        }
        prev = &nums[i]
    }
    log.Printf("increases: %d", increases)
}

func part2(path string) {
    log.SetPrefix("part2: ")
    lines, err := getLines(path)
    check(err)

    nums := make([]int, 0, len(lines))
    for _, line := range lines {
        num, err := strconv.Atoi(line)
        check(err)
        nums = append(nums, num)
    }
    increases := 0
    for i, _ := range nums {
        if i > len(nums) - 4 { break }
        window := nums[i:i + 4]
        if len(window) < 4 { break }
        if window[0] < window[3] { increases++ }
    }
    log.Printf("increases: %d", increases)
}
